src/write/elf/writer.rs

```diff
@@ -1735,6 +1745,75 @@ impl<'a> Writer<'a> {
         });
     }
 
+    /// Reserve the section index for the `.gnu.attributes` section.
+    pub fn reserve_gnu_attributes_section_index(&mut self) -> SectionIndex {
+        debug_assert!(self.gnu_attributes_str_id.is_none());
+        self.gnu_attributes_str_id = Some(self.add_section_name(&b".gnu.attributes"[..]));
+        self.reserve_section_index()
+    }
+
+    /// Reserve the range for the `.gnu.attributes` section.
+    pub fn reserve_gnu_attributes(&mut self, gnu_attributes_size: usize) {
+        debug_assert_eq!(self.gnu_attributes_offset, 0);
+        if gnu_attributes_size == 0 {
+            return;
+        }
+        self.gnu_attributes_size = gnu_attributes_size;
+        self.gnu_attributes_offset = self.reserve(self.gnu_attributes_size, self.elf_align);
+        self.gnu_attributes_size_remaining = self.gnu_attributes_size;
+    }
+
+    /// Write the section header for the `.gnu.attributes` section.
+    ///
+    /// This function does nothing if the section index was not reserved.
+    pub fn write_gnu_attributes_section_header(&mut self, sh_addr: u64) {
+        if self.gnu_attributes_str_id.is_none() {
+            return;
+        }
+        self.write_section_header(&SectionHeader {
+            name: self.gnu_attributes_str_id,
+            sh_type: elf::SHT_GNU_ATTRIBUTES,
+            sh_flags: 0,
+            sh_addr,
+            sh_offset: self.gnu_attributes_offset as u64,
+            sh_size: self.gnu_attributes_size as u64,
+            sh_link: self.dynstr_index.0,
+            sh_info: 0, // TODO
+            sh_addralign: self.elf_align as u64,
+            sh_entsize: 0,
+        });
+    }
+
+    /// Write alignment padding bytes prior to a `.gnu.attributes` section.
+    pub fn write_align_gnu_attributes(&mut self) {
+        if self.gnu_attributes_offset == 0 {
+            return;
+        }
+        util::write_align(self.buffer, self.elf_align);
+        debug_assert_eq!(self.gnu_attributes_offset, self.buffer.len());
+    }
+
+    /// Begin a `.gnu.attributes` section
+    pub fn write_gnu_attributes_version(&mut self, version: u8) {
+        self.buffer.write(&version);
+    }
+
+    /// Begin a `.gnu.attributes` sub-section
+    pub fn write_gnu_attributes_subsection(&mut self, section_length: u32, vendor_name: &[u8]) {
+        self.buffer.write(&U32::new(self.endian, section_length));
+        self.buffer.write_slice(vendor_name);
+        // Write a null byte to end the vendor name
+        self.buffer.write_slice(&[0 as u8]);
+    }
+
+    /// Write a `.gnu.attributes` tag and attribute data
+    pub fn write_gnu_attributes_tag(&mut self, tag: u8, attributes: &[u8]) {
```

Comment:
```suggestion
    pub fn write_gnu_attributes_subsubsection(&mut self, tag: u8, attributes: &[u8]) {
```

Reply:
Since the terminology of this ELF format is a bit of a mess (there's the "gnu.attributes" section, then there are vendor "subsections", but those have their own attribute "subsections") I went with the name ` write_gnu_attributes_subsection` for the vendor subsection, and then `write_gnu_attributes_subsection_attributes` for writing the attribute data within a vendor subsection. I think that's clear enough but let me know if you have a better strategy

Reply:
I agree that the terminology is a bit of a mess, but I chose subsubsection based on the terminology used the documentation you gave. For an example of someone else using this terminology see https://github.com/eliben/pyelftools/blob/master/elftools/elf/sections.py. So I prefer if we use the suggested name of `write_gnu_attributes_subsubsection`.

Reply:
Ah okay I see what you mean now for this, I got dyslexic and missed the second "sub" in "subsub". I think now all the naming should be consistent.